### PR TITLE
Participant metering documentation

### PR DESCRIPTION
**Pull request by simonmaxen-da**
changelog_begin
Participant metering added to Daml operating insturctions documentation
changelog_end

### CODE REVIEW

**Review by resmela-da — CHANGES_REQUESTED** [review]

File: docs/source/ops/metering.rst [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+:participant:
+    The name of the local participant the report applies to
+
+:request:
+    This section gives details of the parameters that were used to generate the report
+
+:final:
+    This field will be set to ``true`` if a from date was provided and the from date is
+    in the past.  Once a report is marked as final the event counts will never change so
+    may be used for billing purposes.
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst:89]: Good spot, `from` should be `to` !

File: docs/source/ops/metering.rst [hunk 2]
```diff
@@ -0,0 +1,96 @@
+.. Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+Participant Metering
+====================
+
+Participant metering is a way to report how many events have been submitted in a given period of time.
+
+This may form part of the commercial arrangement Digital Asset have with customers.
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst]: Done.

File: docs/source/ops/metering.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+A metering report is generated using the :doc:`Daml assistant </tools/assistant>` utility.
+
+To run a metering report ``daml ledger metering-report`` is used with the following metering specific arguments:
+
+:--from:
+    The period being reported on will start from the given date. Events on or after this date will be included.
+
+:--to:
+    An end date that may be used to terminate the reporting period.  Events prior to this date will be included.
+    If an end date is not provided then the report will contains counts of all events that occurred after the to
```
[the remark this answers is not in the document]
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst:37]: Done

File: docs/source/ops/connect/index.rst [hunk 4]
```diff
@@ -3,8 +3,8 @@
 
 .. _ops-connect_index:
 
-Operating Daml
-==============
+Operating Daml Connect
+======================
```
Comment (jwhyte-da) [inline docs/source/ops/connect/index.rst:7]: Why Daml Connect? We don't use this name anymore

Comment (simonmaxen-da) [reply docs/source/ops/connect/index.rst:7]: Mainly because it is in the `connect` sub directory.  As both pages that are contained within the `connect` sub folder (auth0 and helm) are both early access features we could do the following:
- Change the name of the directory to `early-access`
- Change the title in `index.rst` from `Operating Daml Connect` to `Early Access Features`
- In `operating-daml.rst` add the `ops/early-access/index` to the bottom (text displayed will be `Early Access Features`)

File: docs/source/operating-daml.rst [hunk 5]
```diff
@@ -2,9 +2,12 @@
...
 Operating Daml
-===============
+==============
 
 .. toctree::
   :maxdepth: 2
 
-  ops/pruning
\ No newline at end of file
+  ops/connect/index
```
Comment (stefanobaghino-da) [inline docs/source/operating-daml.rst:10]: This line has been removed on purpose in a previous PR. Please remove it.

File: docs/source/ops/connect/index.rst (hunk 4 above)
Comment (stefanobaghino-da) [reply docs/source/ops/connect/index.rst:7]: This has been remove on purpose in a previous PR due to the renaming of the product. Please undo it.

**Review by stefanobaghino-da — CHANGES_REQUESTED** [review]

File: docs/source/ops/connect/index.rst (hunk 4 above)
Comment (stefanobaghino-da) [reply docs/source/ops/connect/index.rst:7]: https://github.com/digital-asset/daml/pull/12668

File: docs/source/operating-daml.rst (hunk 5 above)
Comment (stefanobaghino-da) [reply docs/source/operating-daml.rst:10]: https://github.com/digital-asset/daml/pull/13052

File: docs/source/index.rst [hunk 6]
```diff
@@ -56,7 +56,6 @@ Daml Documentation
    :caption: Operating Daml
 
    operating-daml
-   ops/requirements
```
Comment (stefanobaghino-da) [inline docs/source/index.rst:59]: Why was this line removed?

File: docs/source/ops/metering.rst [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+Generating a Metering Report
+----------------------------
+
+A metering report is generated using the :doc:`Daml assistant </tools/assistant>` utility.
+
+To run a metering report ``daml ledger metering-report`` is used with the following metering specific arguments:
+
+:--from:
+    The period being reported on will start from the given date. Events on or after this date will be included.
```
Comment (carrielaben-da) [inline docs/source/ops/metering.rst:33]:
```suggestion
    A start date that is used to initiate the reporting period. Events on or after this date will be included.
```
To create parallel construction with the end date description.

File: docs/source/ops/connect/index.rst (hunk 4 above)
Comment (stefanobaghino-da) [reply docs/source/ops/connect/index.rst:7]: > Mainly because it is in the `connect` sub directory. As both pages that are contained within the `connect` sub folder (auth0 and helm) are both early access features we could do the following:
> 
> * Change the name of the directory to `early-access`
> * Change the title in `index.rst` from `Operating Daml Connect` to `Early Access Features`
> * In `operating-daml.rst` add the `ops/early-access/index` to the bottom (text displayed will be `Early Access Features`)

There's no need to make any change here, the change has been done purposefully to make sure links wouldn't break across versions. No need to change this.

File: docs/source/ops/metering.rst [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+
+:--to:
+    An end date that may be used to terminate the reporting period.  Events prior to this date will be included.
+    If an end date is not provided then the report will contain counts of all events that occurred after the to
```
Comment (carrielaben-da) [inline docs/source/ops/metering.rst:37]: Do you mean "after the from date"?

File: docs/source/ops/metering.rst [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    date.
+
+:--application:
+    An optional application can be provided to limit the report to a single application.
```
Comment (carrielaben-da) [inline docs/source/ops/metering.rst:41]:
```suggestion
    Optionally, provide an application to limit the report to that application.
```

File: docs/source/ops/metering.rst [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+
+The from and to dates above should be provided formatted as ``yyyy-mm-dd``.  The exact timestamp used for the report
```
Comment (carrielaben-da) [inline docs/source/ops/metering.rst:43]:
```suggestion
The from and to dates above should be formatted ``yyyy-mm-dd``.  The exact timestamp used for the report
```

**Review by cocreature — COMMENTED** [review]
I think you’re trying to do too much here. I think the actual changes are an improvement but I’d suggest you revert anything else and if you want to change the toc or other stuff, open a separate PR for that.

File: docs/source/ops/connect/index.rst (hunk 4 above)
Comment (simonmaxen-da) [reply docs/source/ops/connect/index.rst:7]: Reverted.

File: docs/source/ops/metering.rst (hunk 7 above)
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst:33]: Done.

File: docs/source/ops/metering.rst (hunk 9 above)
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst:41]: Done.

File: docs/source/ops/metering.rst (hunk 10 above)
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst:43]: Done.

File: docs/source/operating-daml.rst (hunk 5 above)
Comment (simonmaxen-da) [reply docs/source/operating-daml.rst:10]: Removed.

File: docs/source/ops/metering.rst (hunk 8 above)
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst:37]: Yes.

File: docs/source/index.rst (hunk 6 above)
Comment (simonmaxen-da) [reply docs/source/index.rst:59]: I moved it to `operating-daml.rst`, bad idea ?

Comment (stefanobaghino-da) [reply docs/source/index.rst:59]: I don't think it's a bad idea, thanks for clarifying.

Comment (cocreature) [inline docs/source/index.rst:59]: Please take a look at https://github.com/digital-asset/daml/blob/main/docs/README.md#updating-the-table-of-contents. We need to mirror this change for the pdf index as well as the index files in the assembly repo.

File: docs/source/ops/metering.rst [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+The events included in the report include:
+
+    *  Contract creation
+    *  Consuming exercise by contract identifier
+    *  Non-consuming exercise by contract identifier
+    *  Consuming contract exercise by contract key
+    *  Non-consuming contract exercise by contract key
+    *  Contract fetch by contract identifier
+    *  Contract fetch by contract key
+    *  Contract lookup by contract key
```
Comment (cocreature) [inline docs/source/ops/metering.rst:20]:
```suggestion
    *  Contract creation
    *  Exercise of a contract (including non-consuming exercises and exercise by key)
     * Fetch of a contract (including fetch by key)
     * Lookup by contract key
```
I think that makes it a bit clearer that this just corresponds to the four type of transaction nodes we have.

File: docs/source/ops/metering.rst [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+
+Note that only events that originated from the local participant are included in the metering.  Events received
```
Comment (cocreature) [inline docs/source/ops/metering.rst:22]: It might also be worth pointing out exlicitly that only a successful, committed transaction is metered not one that fails. Saying you meter tranascitons rather than commands points that out implicitly but I think it’s worth making explicit.

File: docs/source/ops/metering.rst [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+by the local participant from remote participants are *not* included.
+
+Generating a Metering Report
+----------------------------
+
+A metering report is generated using the :doc:`Daml assistant </tools/assistant>` utility.
+
+To run a metering report ``daml ledger metering-report`` is used with the following metering specific arguments:
+
+:--from:
```
Comment (cocreature) [inline docs/source/ops/metering.rst:32]:
```suggestion
:``--from``:
```

Comment (cocreature) [reply docs/source/ops/metering.rst:32]: otherwise the double dashes are not rendered as double dashes.

File: docs/source/ops/metering.rst [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    A start date that is used to initiate the reporting period. Events on or after this date will be included.
+
+:--to:
```
Comment (cocreature) [inline docs/source/ops/metering.rst:35]:
```suggestion
:``--to``:
```

File: docs/source/ops/metering.rst [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+    An end date that may be used to terminate the reporting period.  Events prior to this date will be included.
+    If an end date is not provided then the report will contain counts of all events that occurred on or after
+    the ``--from`` date.
+
+:--application:
```
Comment (cocreature) [inline docs/source/ops/metering.rst:40]:
```suggestion
:``--application``:
```

File: docs/source/ops/metering.rst (hunk 12 above)
Comment (simonmaxen-da) [reply docs/source/ops/metering.rst:18]: Done.

File: docs/source/index.rst (hunk 6 above)
Comment (simonmaxen-da) [reply docs/source/index.rst:59]: `docs/configs/pdf/index.rst` updted.

Comment (cocreature) [reply docs/source/index.rst:59]: I think you forgot to push.

**Review by cocreature — APPROVED** [review]
nice work thanks! I think you forgot to push the changes that address my 2 remaining comments.

File: docs/source/ops/metering.rst [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    * Contract creation
+    * Exercise of a contract (including non-consuming exercises and exercise by key)
+    * Fetch of a contract (including fetch by key)
+    * Lookup by contract key
+
+Only events that originated from the local participant are included in the metering.  Events received
+by the local participant from remote participants are *not* included.
+
+Only events contained in an accepted transactions (ones that maintains the
+:doc:`integrity of the ledger model </concepts/ledger-model/ledger-integrity>`) are included.
```
Comment (cocreature) [inline docs/source/ops/metering.rst:22]:
```suggestion
Only events contained in committed transactions are included, a failed transaction has no effect on ledger metering.
```
